File: internal/moleman/template_test.go
package moleman

import "testing"

func TestRenderTemplate(t *testing.T) {
	data := map[string]any{
		"input": map[string]any{
			"prompt": "hello",
		},
		"outputs": map[string]any{
			"review_json": map[string]any{
				"structured_output": map[string]any{
					"must_fix_count": 0,
				},
			},
		},
	}

	out, err := RenderTemplate("{{ .input.prompt }} {{ .outputs.review_json.structured_output.must_fix_count }}", data)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if out != "hello 0" {
		t.Fatalf("unexpected output: %q", out)
	}
}

func TestRenderTemplateShellEscape(t *testing.T) {
	data := map[string]any{
		"input": map[string]any{
			"prompt": "we're good",
		},
	}

	out, err := RenderTemplate("{{ shellEscape .input.prompt }}", data)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if out != "'we'\"'\"'re good'" {
		t.Fatalf("unexpected output: %q", out)
	}
}

func TestRenderTemplateEmptyInput(t *testing.T) {
	out, err := RenderTemplate("", map[string]any{"ignored": true})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if out != "" {
		t.Fatalf("unexpected output: %q", out)
	}
}

File: test/template.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import { renderTemplate, shellEscape } from "../src/template.js";

test("renderTemplate supports value interpolation", () => {
  const data = {
    input: {
      prompt: "hello",
    },
    outputs: {
      review_json: {
        structured_output: {
          must_fix_count: 0,
        },
      },
    },
  };

  const out = renderTemplate("{{ .input.prompt }} {{ .outputs.review_json.structured_output.must_fix_count }}", data);
  assert.equal(out, "hello 0");
});

test("renderTemplate supports index helper", () => {
  const data = {
    outputs: {
      fix: "done",
    },
  };
  const out = renderTemplate("{{ index .outputs \"fix\" }}", data);
  assert.equal(out, "done");
});

test("shellEscape matches expected format", () => {
  assert.equal(shellEscape("we're good"), "'we'\"'\"'re good'");
});

test("renderTemplate handles empty input", () => {
  const out = renderTemplate("", { ignored: true });
  assert.equal(out, "");
});
